### PR TITLE
Add @embedder decorator to exported functions

### PR DESCRIPTION
**Pull request by JairusSW**
This PR adds support for the `@embedder` decorator

Completes HYP-1395

### CODE REVIEW

File: src/transform/src/metadata.ts [hunk 1]
```diff
@@ -40,6 +41,24 @@ export class HypermodeMetadata {
...
 
+  addEmbedders(functions: FunctionSignature[]) {
+    for (const func of functions) {
+      if (func.parameters.length != 1)
+        throw new Error(
+          "Expected embedder to have one argument, but found " +
+            (func.parameters.length ? "multiple" : "no") +
+            " arguments!",
+        );
+      if (func.parameters[0].type.name != "string")
```
Comment (jairad26) [inline src/transform/src/metadata.ts:52]: to future proof this, could you change it to support string[] as well, we want to allow batching. And separately, can it also check that the return type is f64[] or f64[][]

Comment (mattjohnsonpint) [reply src/transform/src/metadata.ts:52]: Yeah, let's support two signatures.  Either `string => f64[]` or `string[] => f64[][]`

File: src/transform/src/extractor.ts [hunk 2]
```diff
@@ -16,23 +16,48 @@ import {
...
     this.program = transform.program;
     this.binaryen = transform.binaryen;
     this.module = module;
+    this.transform = transform;
   }
 
   getProgramInfo(): ProgramInfo {
+    if (!(this.transform instanceof HypermodeTransform)) {
+      console.warn(
+        "Using custom transform. Cannot use @embedded unless using the HypermodeTransform!",
```
Comment (jairad26) [inline src/transform/src/extractor.ts:37]: should be "@embedder" in the warn

File: src/transform/src/metadata.ts (hunk 1 above)
Comment (JairusSW) [reply src/transform/src/metadata.ts:52]: Sure

File: src/transform/src/index.ts [hunk 3]
```diff
@@ -3,14 +3,43 @@ import { createWriteStream } from "fs";
...
+  NodeKind,
+  CommonFlags,
+} from "assemblyscript/dist/assemblyscript.js";
+import { isStdlib } from "./utils.js";
 
 export default class HypermodeTransform extends Transform {
+  public embedders: string[] | null = null;
+  afterParse(parser: Parser): void | Promise<void> {
+    for (const source of parser.sources) {
+      if (source.isLibrary) continue;
```
Comment (mattjohnsonpint) [inline src/transform/src/index.ts:19]: Would this prevent us from making a library of embedders, or including some default embedders in the existing `@hypermode/functions-as` library?

File: src/transform/src/metadata.ts [hunk 4]
```diff
@@ -117,6 +136,10 @@ export class HypermodeMetadata {
     this.functions.forEach((f) => writeItem(f.toString()));
     stream.write("\n");
 
+    writeHeader("Hypermode Embedders:");
```
Comment (mattjohnsonpint) [inline src/transform/src/metadata.ts:139]:
```suggestion
    writeHeader("Embedder Functions:");
```

(just a branding thing)

File: src/transform/src/extractor.ts [hunk 5]
```diff
@@ -16,23 +16,48 @@ import {
...
 
   constructor(transform: Transform, module: binaryen.Module) {
     this.program = transform.program;
     this.binaryen = transform.binaryen;
     this.module = module;
+    this.transform = transform;
   }
 
   getProgramInfo(): ProgramInfo {
+    if (!(this.transform instanceof HypermodeTransform)) {
```
Comment (mattjohnsonpint) [inline src/transform/src/extractor.ts:35]: We don't need to support other transforms, so the scenario being checked isn't likely.

File: src/transform/src/extractor.ts [hunk 6]
```diff
@@ -16,23 +16,48 @@ import {
   TypeInfo,
   typeMap,
 } from "./types.js";
+import HypermodeTransform from "./index.js";
 
 export class Extractor {
   binaryen: typeof binaryen;
   module: binaryen.Module;
   program: Program;
+  transform: Transform;
```
Comment (mattjohnsonpint) [inline src/transform/src/extractor.ts:25]:
```suggestion
  transform: HypermodeTransform;
```
Then you don't need to check it later.

File: src/transform/src/metadata.ts (hunk 1 above)
Comment (JairusSW) [reply src/transform/src/metadata.ts]: @jairad26 @mattjohnsonpint resolved.

File: src/transform/src/index.ts (hunk 3 above)
Comment (JairusSW) [reply src/transform/src/index.ts:19]: You can only export from the entry file. It *should* handle cases like `export { embedderFunction } from "../path-to-embedder.ts";`
I should check that

Comment (JairusSW) [reply src/transform/src/index.ts:19]: Okay, it doesn't work. Adding support now

Comment (JairusSW) [reply src/transform/src/index.ts]: Done.

**Comment by JairusSW** [comment]
If there are no embedders, should it print like `Embedder Functions: None` or just be `Embedder Functions:`

**Review by mattjohnsonpint — APPROVED** [review]

**Pull request merged**